referencing interface object causes c error: pointer expected

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 535b04f, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 5c65e58.535b04f
|:-------------------|:-------------------
|OS                  |windows, Microsoft Windows 11 Pro 26100 64-bit
|Processor           |24 cpus, 64bit, little endian, AMD Ryzen 9 3900X 12-Core Processor  
|Memory              |20.53GB/31.9GB
|                    |
|V executable        |C:\v\v.exe
|V last modified time|2025-02-02 09:32:13
|                    |
|V home dir          |OK, value: C:\v
|VMODULES            |OK, value: C:\Users\info\.vmodules
|VTMP                |OK, value: C:\Users\info\AppData\Local\Temp\v_0
|Current working dir |OK, value: D:\vbug2
|                    |
|Git version         |git version 2.42.0.windows.2
|V git status        |weekly.2025.1-155-g535b04f4-dirty
|.git/config present |true
|                    |
|cc version          |N/A
|gcc version         |N/A
|clang version       |clang version 17.0.1
|msvc version        |N/A
|tcc version         |tcc version 0.9.27 (x86_64 Windows)
|tcc git status      |thirdparty-windows-amd64 b425ac82
|emcc version        |N/A
|glibc version       |ldd (cygwin) 3.4.7

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  main.v && D:\vbug2\main.exe`
```v
module main

@[heap]
interface IGameObject {
mut:
	name     string
	parent   ?&IGameObject
	children []&IGameObject
	add_child(mut o IGameObject)
}

@[heap]
struct GameObject implements IGameObject {
mut:
	name     string
	parent   ?&IGameObject
	children []&IGameObject
}

fn (mut gameobject GameObject) add_child(mut o IGameObject) {
	o.parent = gameobject
	gameobject.children << o
}

fn main() {
	mut v1 := &GameObject{
		name: 'v1'
	}
	mut v2 := &GameObject{
		name: 'v2'
	}
	v1.add_child(mut v2)
	for child in v1.children {
		if child.parent != none {
			eprintln('parent: ${child.parent.name}')
		}
	}
}

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: C:/Users/info/AppData/Local/Temp/v_0/main.01JK329G1WXGW8DEKSJER12Z08.tmp.c:4117: warning: implicit declaration of function 'tcc_backtrace'
cc: C:/Users/info/AppData/Local/Temp/v_0/main.01JK329G1WXGW8DEKSJER12Z08.tmp.c:6898: error: pointer expected
... (the original output was 3 lines long, and was truncated to 2 lines)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

no c error



> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4012,10 +4012,11 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	mut sum_type_dot := '.'
 	mut field_typ := ast.void_type
 	mut is_option_unwrap := false
+	is_iface_or_sumtype := sym.kind in [.interface, .sum_type]
 	if f := g.table.find_field_with_embeds(sym, node.field_name) {
 		field_sym := g.table.sym(f.typ)
 		field_typ = f.typ
-		if sym.kind in [.interface, .sum_type] {
+		if is_iface_or_sumtype {
 			g.write('(*(')
 		}
 		is_option := field_typ.has_flag(.option)
@@ -4036,7 +4037,7 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 					for i, typ in field.smartcasts {
 						if i == 0 && (is_option_unwrap || nested_unwrap) {
 							deref := if g.inside_selector {
-								if sym.kind in [.interface, .sum_type] {
+								if is_iface_or_sumtype {
 									'*'.repeat(field.smartcasts.last().nr_muls())
 								} else {
 									'*'.repeat(field.smartcasts.last().nr_muls() + 1)
@@ -4142,7 +4143,7 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 			return
 		}
 	} else {
-		if sym.kind in [.interface, .sum_type] {
+		if is_iface_or_sumtype {
 			g.write('(*(')
 		}
 	}

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4036,7 +4036,11 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 					for i, typ in field.smartcasts {
 						if i == 0 && (is_option_unwrap || nested_unwrap) {
 							deref := if g.inside_selector {
-								'*'.repeat(field.smartcasts.last().nr_muls() + 1)
+								if sym.kind in [.interface, .sum_type] {
+									'*'.repeat(field.smartcasts.last().nr_muls())
+								} else {
+									'*'.repeat(field.smartcasts.last().nr_muls() + 1)
+								}
 							} else if sym.kind == .interface && !typ.is_ptr()
 								&& field.orig_type.has_flag(.option) {
 								''

--- a/vlib/v/tests/options/option_ptr_unwrap_selector_test.v
+++ b/vlib/v/tests/options/option_ptr_unwrap_selector_test.v
@@ -1,0 +1,39 @@
+module main
+
+@[heap]
+interface IGameObject {
+mut:
+	name     string
+	parent   ?&IGameObject
+	children []&IGameObject
+	add_child(mut o IGameObject)
+}
+
+@[heap]
+struct GameObject implements IGameObject {
+mut:
+	name     string
+	parent   ?&IGameObject
+	children []&IGameObject
+}
+
+fn (mut gameobject GameObject) add_child(mut o IGameObject) {
+	o.parent = gameobject
+	gameobject.children << o
+}
+
+fn test_main() {
+	mut v1 := &GameObject{
+		name: 'v1'
+	}
+	mut v2 := &GameObject{
+		name: 'v2'
+	}
+	v1.add_child(mut v2)
+	for child in v1.children {
+		if child.parent != none {
+			eprintln('parent: ${child.parent.name}')
+		}
+	}
+	assert true
+}

